Library Rewrite

Luc1412 opened the pull request:
# Library Rewrite
This PR encompasses the Version 3.0 bump of the Fortnite API Python wrapper. This major transition gives the library a complete facelift, focusing on asynchronous functionality while still supporting the synchronous library you've come to know.

~~To see how progress is coming along, it's best to check out the [Library Rewrite Project](https://github.com/orgs/Fortnite-API/projects/1/views/2) and the rewrite's [unit tests](https://github.com/Fortnite-API/py-wrapper/tree/rewrite/tests) for the library, as these are the most up-to-date at the moment.~~

(*Edited May 4th, 2024*): Progress is coming along fantastic. As we prepare to close out this project, documentation has been the main focus for our users. A complete migration guide has been built from the ground up to ensure a clean transition from Version 2 of the library to the new Version 3. ~~Here's a sneak peek:~~

All that remains to be completed is working through every aspect of the documentation with a fine-tooth-comb for any improvements in clarity and consistency. The core API should not change much anymore. 

(*Edited May 7th, 2024*): Documentation is posted and can be found here: https://fortnite-api.readthedocs.io/en/rewrite/

(*Edited Aug 4th, 2024*): Welp, last time I spoke on May 4th I lied. Life happens, what can I say. The project has been completed... at least the code base... actually this time. All that remains is:
- [x] Opt in data fields, pending API update for the Python implementation
- [x] One unresolved discussion regarding helper functions in the POI class.
- [x] Going through all the codebase with a fine-tooth comb for a last review
- [x] Going through all the documentation and migration guide for any needed small changes.

I will update this check list as they get completed. Cheers 🍺 

## Features and Additions
- Complete overhaul of every object, method, and property in the library. Now, fetching information from the API is done through methods on a ~~`FortniteAPI`~~`Client` and ~~`SyncFortniteAPI`~~`SyncClient` class. This is to make the library easier to use and maintain, while still being welcoming for new programmers.
- Complete overhaul of documentation. This includes a major shift to readthedocs powered by Sphinx for easier documentation.
- Speed improvements for large endpoints, such as fetching all available game cosmetics at once.
- Strictly type hinted with high code standards.

Luc1412 commented on fortnite_api/cosmetics/common.py:
```diff
@@ -0,0 +1,176 @@
+        self.lego: Optional[Asset[HTTPClientT]] = lego and Asset(http=http, url=lego)
+
+        # Mapping of str to Asset. In testing all I have found is "background" in this mapping.
+        other: Dict[str, str] = get_with_fallback(data, 'other', dict)
```
We can be more specific here. Images include:
- background for loading screens
- coverart for lobby music
- decal for sprays

trevorflahardy replied:
By any chance can you provide the specific keys for this?

Luc1412 replied:
`background`, `coverart` and `decal` are the keys

trevorflahardy replied:
Should this remain under the `other` attribute or become attributes of the CosmeticImages object itself?

Luc1412 replied:
I think keeping other might be unnecessary nesting. So maybe we keep it in the CosmeticImages.

Luc1412 commented on fortnite_api/news.py:
```diff
@@ -16,39 +52,59 @@ class News:
+        self.stw: Optional[GameModeNews[HTTPClientT]] = _stw and GameModeNews(data=_stw, http=http)
 
+        _creative = data.get('creative')
+        self.creative: Optional[GameModeNews[HTTPClientT]] = _creative and GameModeNews(data=_creative, http=http)
```
We probably can remove this, since creative news have been discontinued since later 2021. This field just exists as legacy

Luc1412 commented on fortnite_api/stats.py:
```diff
@@ -1,65 +1,261 @@
+        self.duo: Optional[BrGameModeStats] = _duo and BrGameModeStats(data=_duo)
+
+        _trio = data.get('trio')
+        self.trio: Optional[BrGameModeStats] = _trio and BrGameModeStats(data=_trio)
```
We probably can remove this, since trio news have been removed in favour of representing stats as it is in-game. This field is always empty and just exists as a legacy.

Luc1412 commented on fortnite_api/stats.py:
```diff
@@ -1,65 +1,261 @@
+        The score per match for the stats for a specific game mode. This is the score divided by the matches played.
+    wins: :class:`int`
+        The total number of wins in this specific game mode.
+    top3: :class:`int`
+        The number of times the player has placed in the top 3 in this specific game mode.
+    top5: :class:`int`
+        The number of times the player has placed in the top 5 in this specific game mode.
+    top6: :class:`int`
+        The number of times the player has placed in the top 6 in this specific game mode.
+    top10: :class:`int`
+        The number of times the player has placed in the top 10 in this specific game mode.
+    top12: :class:`int`
+        The number of times the player has placed in the top 12 in this specific game mode.
+    top25: :class:`int`
+        The number of times the player has placed in the top 25 in this specific game mode.
```
We can be more specific in the docs. Specific fields are only set for specific team sizes:
### 10 players remaining:
- top10 = Solo
- top5 = Duo
- top3 = Squad
### 25 players remaining:
- top25 = Solo
- top12 = Duo
- top6 = Squad

trevorflahardy replied:
Updated in [801d6ed](https://github.com/Fortnite-API/py-wrapper/pull/19/commits/801d6edb6c26c2416798561752b2a1137ea6672f) - this appears to have happened because of a major oversight due to no stat tests. Adding stat tests now.

Luc1412 commented on fortnite_api/cosmetics/instrument.py:
```diff
@@ -0,0 +1,103 @@
+        The gameplay tags of the instrument.
+    path: Optional[:class:`str`]
+        The path of the instrument.
+    showcase_video: Optional[:class:`str`]
+        The showcase video of the instrument.
```
Since it's a Youtube Video ID, we should be more precise. We could rename it to `showcase_video_id` and mention in the docs it's a YoutTube URL or to `showcase_video_url` and provide the full youtube url.

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,272 @@
+    Attributes
+    ----------
+    value: :class:`str`
+        The content of the banner.
```
I think "The content of the banner." or "The text of the banner." might be better here

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,272 @@
+
+    def __init__(self, *, data: Dict[str, Any]) -> None:
         self.value: str = data['value']
         self.intensity: str = data['intensity']
```
We can create an Enum. Values are Low and High

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,272 @@
+        self.dev_name: str = data['devName']
+        self.offer_id: str = data['offerId']
+        self.display_asset_path: Optional[str] = data.get('displayAssetPath')
+        self.tile_size: str = data['tileSize']
```
We can Create an enum for TileSize. Values are: Small, Normal, DoubleWide and TripleWide. There was also a case, where Epic got a type in the tile size and it was spelled "nomal" instead of normal. In-game it was shown properly as Normal width, so this might be the fallback.

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,272 @@
+    index: :class:`int`
+        The index of the layout.
+    show_ineligible_offers: :class:`str`
+        The show ineligible offers flag of the layout.
```
I think this need to be better worded. Something like `Whether ineligible offers are displayed in the layout`

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,272 @@
+    value: :class:`str`
+        The content of the banner.
+    intensity: :class:`str`
+        The visual style or intensity of the banner.
```
Maybe we keep it just as `The intensity of the banner.`

Luc1412 commented on fortnite_api/__init__.py:
```diff
@@ -1,16 +1,44 @@
 from .playlist import *
 from .shop import *
 from .stats import *
+from .utils import *
```
Do we want to expose all utils?

trevorflahardy replied:
Adjusted

trevorflahardy committed:
Fix how playlists are checked during tests

trevorflahardy committed:
Move major classes to new proxy type

trevorflahardy committed:
Fix iterator import

trevorflahardy committed:
Update setup.py

trevorflahardy committed:
Refactor GH workflows

trevorflahardy committed:
Update workflows structure

trevorflahardy committed:
Update requirements.txt for typing extensions

trevorflahardy committed:
Correct docs requires and update conf for auto enum

trevorflahardy committed:
Add speed extras

trevorflahardy committed:
Updates

trevorflahardy committed:
Cleanups as I go

trevorflahardy committed:
Start adding some base documentation

trevorflahardy committed:
Updates

trevorflahardy committed:
Add all clases to docs, need to fix links later

trevorflahardy committed:
Black isort

trevorflahardy committed:
Update documentation references, start building with no errors

trevorflahardy committed:
Black isort

trevorflahardy committed:
Fix header length

trevorflahardy committed:
Correct document attributes on client classes

trevorflahardy committed:
Add exception hierarchy extension

trevorflahardy committed:
Fix some documentation errors

trevorflahardy committed:
Add some notes to exception hierarchy

trevorflahardy committed:
Update typehints

trevorflahardy committed:
Update type hints for exception hierarchy

trevorflahardy committed:
Add exception hierarchy

Luc1412 committed:
Update TileSize values

trevorflahardy committed:
Updates to errors

trevorflahardy committed:
Update errors

trevorflahardy committed:
Add some ratelimit handling

trevorflahardy committed:
Add now func and update how datetimes are parsed

trevorflahardy committed:
Add tests for ratelimits

trevorflahardy committed:
Correct for tile size update

trevorflahardy committed:
Update tests dep

trevorflahardy committed:
Black formatting for whitespace

trevorflahardy committed:
Merge branch 'master' into rewrite

trevorflahardy committed:
Start adding migrating documentation

trevorflahardy committed:
Update migrating.rst

trevorflahardy committed:
Update default flags

trevorflahardy committed:
Make move to decorator for beta funcs, and update docs and flags

trevorflahardy committed:
Update tests for these changes

trevorflahardy committed:
Continue working on migration documentation

trevorflahardy committed:
Add a workaround for TileSize being improper sometimes

trevorflahardy committed:
Fix from value fallback, it was more than expected

trevorflahardy committed:
Black isort formatting

trevorflahardy committed:
Fix doc issue

trevorflahardy committed:
Add some inherited members

trevorflahardy committed:
Add docs

trevorflahardy committed:
Fix typo issue

trevorflahardy committed:
Add license on conftest

trevorflahardy committed:
Add outdated code block ext

trevorflahardy committed:
Add css for outdated code blocks

trevorflahardy committed:
Move to outdated code blocks

trevorflahardy committed:
Fix nitpicky configuration in conf.py

trevorflahardy committed:
Update license headers in exception_hierarchy.py and outdated_code_blocks.py

trevorflahardy committed:
Continue working on migrations

Luc1412 committed:
Add texture to MaterialInstanceImages

trevorflahardy committed:
Add texture to slots

trevorflahardy committed:
Move from TileSize enum to class, update docs for the change and fix tests

trevorflahardy committed:
Isort

trevorflahardy committed:
Correct build issue for options

trevorflahardy committed:
Remove locale dir pointer

trevorflahardy committed:
Add a readme to the docs dir so I don't forget how to build the docs

trevorflahardy committed:
Merge two br stat fetch methods into one

trevorflahardy committed:
Refactor all tests to use api key, update stat tests for updated stat methods

trevorflahardy committed:
Remove alias BannerColour

trevorflahardy committed:
Correct documentation error

trevorflahardy committed:
Add class ref on enum

trevorflahardy committed:
Add major section to documentation, will refine soon

trevorflahardy committed:
Black isort

trevorflahardy committed:
Remove unused imports

trevorflahardy committed:
Fix documentation migration build issues

trevorflahardy committed:
Cleanup how beta exceptions are managed

trevorflahardy committed:
Cleanup documentation and add exception docs

trevorflahardy committed:
Update conf

trevorflahardy committed:
Add logo ico for docs

trevorflahardy committed:
Move to separate api files for dropdown on side bar

trevorflahardy committed:
Update max depth

trevorflahardy committed:
Update examples

trevorflahardy committed:
chore: Add parallel read and write safety for exception hierarchy

trevorflahardy committed:
chore: Add parallel read and write safety for exception hierarchy

trevorflahardy committed:
Update typehint for transformer list proxy in prep for flag

trevorflahardy committed:
Rename param for remove coro doc

trevorflahardy committed:
Correct flag to enum auto

trevorflahardy committed:
Merge beta decorator into one func and add tests for it

trevorflahardy committed:
black isort

trevorflahardy committed:
Correct arguments for sync client

trevorflahardy committed:
Update how shop tile size handles API backups

trevorflahardy committed:
Merge branch 'master' into rewrite

trevorflahardy committed:
Add type aliases for complex func definitions

trevorflahardy committed:
Add attribute table directive

trevorflahardy committed:
Isort on code blocks

trevorflahardy committed:
Add attribute tables onto all classes

trevorflahardy committed:
Add ratelimit note for BR stats fetch

trevorflahardy committed:
Remove box shadow

trevorflahardy committed:
Cleanup docs for Client

trevorflahardy committed:
Update docs for context managers migrating

trevorflahardy committed:
Correct some verbiage, move to MatchMethod on search

trevorflahardy committed:
feat: Add MissingAPIKey exception class

trevorflahardy committed:
Cleanup documentation

trevorflahardy committed:
Update warning for beta method, add code example

trevorflahardy committed:
Update account docs

trevorflahardy committed:
Cleanup AES documentation

trevorflahardy committed:
Cleanup documentation for CosmeticsAll

trevorflahardy committed:
Update Asset documentation for an example and to cleanup docs

trevorflahardy committed:
Update docs for banner

trevorflahardy committed:
Update documentation for creator code

trevorflahardy committed:
Update Images documentation

trevorflahardy committed:
Update docs for map

trevorflahardy committed:
Add flags test and update docs/typehints

trevorflahardy committed:
Add dpy to dev requirements

trevorflahardy committed:
Major transition:

FortniteAPI -> Client
SyncFortniteAPI -> SyncClient

trevorflahardy committed:
Black isort after transition

trevorflahardy committed:
Rename CosmeticType to CosmeticCategory

trevorflahardy committed:
Rename CosmeticBrType to CosmeticType to denote broader purpose of class

trevorflahardy committed:
Update documentation for this change

trevorflahardy committed:
Renaming for standard conventions and update documentation

trevorflahardy committed:
Update documentation for the value change

trevorflahardy committed:
Rename CosmeticSeries to CosmeticSeriesInfo to be more consistent with other classes

trevorflahardy committed:
Mark how someone can get an API key

trevorflahardy committed:
Add docs for client

trevorflahardy committed:
Add installation instructions

trevorflahardy committed:
Update README

trevorflahardy committed:
Add typing metadata

trevorflahardy committed:
Add package data for stubs

trevorflahardy committed:
Correct key error with stats endpoint

trevorflahardy committed:
Fix shop issue for shop layout potentially being None

trevorflahardy committed:
Fix key issue in shop

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+    :since: v2.6.6
+
+    client = fortnite_api.Client(api_key='', run_async=False)
+
```
In V2 it wasn't named client

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+
+Async client
+^^^^^^^^^^^^
+.. list-table::
+    :header-rows: 1
+
+    *   - Parameter
+        - Description
+    *   - ``api_key``
+        - The API key to use for requests to the Fortnite API. This is not a required parameter for most endpoints however, it is required to use any stats endpoints. This is the same as in Version 2. 
+    *   - ``default_language``
+        - The client now boasts a default language parameter. For many endpoints in the Fortnite API, you can specify a language parameter to denote the language you want the response to be in. This parameter allows you to set a default language for all requests made by the client, for endpoints that support it, so that you don't have to specify it every time you make a request. The default langu ...
+    *   - ``session``
+        - The session parameter is a new parameter that allows you to pass a :class:`aiohttp.ClientSession` to the client. By default this is an optional parameter. This parameter comes in useful if you want to manage a session yourself. In that case, it is a required parameter.
+    *   - ``beta``
+        - Denotes if the client has the permissions to access beta endpoints (always prefixed with ``beta_fetch_x``). This is a new parameter that is set to ``False`` by default. If you want to access beta endpoints, you must set this parameter to ``True``.
+    *   - ``optimization_flags``
+        - This is a new parameter that allows you to toggle certain optimizations on or off. This parameter is not required, and defaults to :meth:`~fortnite_api.OptimizationFlags.default` if not given. 
+
+See the documentation for these parameters in :class:`fortnite_api.Client`.
+
+Sync client
+^^^^^^^^^^^
+.. list-table::
+    :header-rows: 1
+
+    *   - Parameter
+        - Description
+    *   - ``api_key``
+        - The API key to use for requests to the Fortnite API. This is not a required parameter for most endpoints however, it is required to use any stats endpoints. This is the same as in Version 2. 
+    *   - ``default_language``
+        - The client now boasts a default language parameter. For many endpoints in the Fortnite API, you can specify a language parameter to denote the language you want the response to be in. This parameter allows you to set a default language for all requests made by the client, for endpoints that support it, so that you don't have to specify it every time you make a request. The default langu ...
+    *   - ``session``
+        - The session parameter is a new parameter that allows you to pass a :class:`requests.Session` to the client. By default this is an optional parameter. This parameter comes in useful if you want to manage a session yourself. In that case, it is a required parameter.
+    *   - ``beta``
+        - Denotes if the client has the permissions to access beta endpoints (always prefixed with ``beta_fetch_x``). This is a new parameter that is set to ``False`` by default. If you want to access beta endpoints, you must set this parameter to ``True``.
+    *   - ``optimization_flags``
+        - This is a new parameter that allows you to toggle certain optimizations on or off. This parameter is not required, and defaults to :meth:`~fortnite_api.OptimizationFlags.default` if not given. 
+
+See the documentation for these parameters in :class:`fortnite_api.SyncClient`.
```
Maybe we can shorten that, since params are almost the same.

trevorflahardy replied:
This specific design choice was made to link new users to the migration documentation. It may be unnecessarily difficult for users to understand the migration guide if they are trying to determine which client they are using and which parameters line up with which client... I feel it's going to lead to silly questions. This clears up any ambiguity - What are your thoughts?

Luc1412 replied:
I thought about starting that both clients params are identical. Only session is a different type, that can be mentioned in the row by splitting it into 2 parts.

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+    :since: v2.6.6
+
+    cosmetics = await client.cosmetics.fetch()
+    reveal_type(cosmetics)
+    >>> List[fortnite_api.Cosmetic]
```
In both cases reveal type might be confusing

trevorflahardy replied:
`print(x)` is better here but every Cosmetic does not implement `repr(x)`. `repr` should probably be implemented and then transform the documentation to `print(...)`, do you agree?

Luc1412 replied:
I think that might be better. Almost no user probably knows `reveal_type ` or some might be even not too much involved in the concept of typing.

trevorflahardy replied:
Fixed with repr(x)

Luc1412 commented on fortnite_api/account.py:
```diff
@@ -1,18 +1,83 @@
+
+        self.external_auths: Dict[Any, Any] = get_with_fallback(
+            data, 'external_auths', dict
+        )  # Adding when User lookup feature is enabled
```
I think we should remove that until it's actually added

trevorflahardy replied:
Agreed, removed.

Luc1412 commented on fortnite_api/aes.py:
```diff
@@ -1,47 +1,195 @@
+    Attributes
+    ----------
+    main_key: Optional[:class:`str`]
+        The main encryption key. Can be ``None`` if Fortnite was recently updated.
```
Can be None if the current versions key isn't available yet.

trevorflahardy replied:
Corrected terminology.

Luc1412 commented on setup.py:
```diff
@@ -5,16 +5,27 @@
+extras_require = {
+    'tests': ['pytest', 'pytest-asyncio', 'pytest-cov', 'python-dotenv', 'pytest-mock'],
+    'docs': ['sphinx', 'sphinxcontrib_trio', 'sphinxcontrib-websupport', 'typing-extensions', 'furo', 'sphinx-copybutton'],
+    'dev': ['black', 'isort', 'discord.py'],
```
Maybe pinning versions makes sense, also isn't pyright missing?

Luc1412 commented on fortnite_api/utils.py:
```diff
@@ -0,0 +1,184 @@
+        return datetime.datetime.strptime(timestamp, '%Y-%m-%dT%H:%M:%S%z')
+
+
+def now() -> datetime.datetime:
```
Maybe `utcnow()`?

trevorflahardy replied:
Not in this case because `utcnow()` returns the UTC date as a naive datetime obj - ie. it doesn't contain any timezone info. It's important to keep the obj aware of timezones for proper ratelimit handling.

Luc1412 replied:
Sorry, my comment was a bit too broad. I meant renaming the function to `utcnow()`. In 3.12 `datetime.utcnow()` is deprecated anyway.

Luc1412 commented on fortnite_api/stats.py:
```diff
@@ -1,65 +1,284 @@
+    players_outlived: :class:`int`
+        The total number of players outlived in this specific game mode.
+    last_modified: :class:`datetime.datetime`
+        The date when this data was last modified.
```
Maybe it makes sense to refer to the last time the player played, even trough the stats might not be updated yet. Not sure.

trevorflahardy replied:
Is the `last_modified` field one that the Epic Games API provides, or is it a datetime that the Fortnite-API updates manually?

If the former, I agree completely - if the latter, however, it may be good to make a note that the field shows the last time the specific stat data was updated.

Luc1412 replied:
> Is the last_modified field one that the Epic Games API provides

Yeah, it is. It shows the date, when the stats where updated within Epics API.

Luc1412 commented on setup.py:
Do we want to move most stuff into `pyproject.toml`?

Luc1412 commented on fortnite_api/enums.py:
```diff
@@ -48,10 +185,46 @@ class BrCosmeticRarity(Enum):
     UNCOMMON = 'uncommon'
     COMMON = 'common'
-    UNKNOWN = 'unknown'
+    MYTHIC = 'mythic'
```
There are no cosmetics with mythic rarity.

Luc1412 commented on fortnite_api/enums.py:
```diff
@@ -90,45 +263,114 @@ class BrCosmeticType(Enum):
+    """
+
+    LOW = 'Low'
+    MEDIUM = 'Medium'
```
As far as I know, medium doesn't exist.

Luc1412 commented on fortnite_api/map.py:
```diff
@@ -1,25 +1,145 @@
+    def __init__(self, *, data: Dict[str, Any], http: HTTPClientT) -> None:
+        self.images: MapImages[HTTPClientT] = MapImages(data=data['images'], http=http)
+
+        self._pois: Mapping[str, POI] = {poi['id']: POI(data=poi) for poi in data['pois']}
+        self.raw_data: Dict[str, Any] = data
+
+    @property
+    def pois(self) -> List[POI]:
+        """List[:class:`fortnite_api.POI`]: The list of POIs in the map."""
+        return list(self._pois.values())
+
+    def get_poi(self, /, id: str) -> Optional[POI]:
+        """Get a POI by its ID.
+
+        Parameters
+        ----------
+        id: :class:`str`
+            The ID of the POI.
+
+        Returns
+        -------
+        Optional[:class:`fortnite_api.POI`]
+            The POI if found, else ``None``.
+        """
+        return self._pois.get(id)
```
Do you believe this is necessary? I think providing a list could be enough.

trevorflahardy replied:
I believe it's necessary and helpful but when reflecting upon it I see a discrepancy. The library does not have any other `get_x` functions, so for this to make sense many other objects in the library would have to follow a similar pattern.

For simplicity I'm going to remove it, and later down the road a new design decision can be made if need be.

Luc1412 commented on fortnite_api/map.py:
```diff
@@ -1,25 +1,145 @@
+        The ID of the POI.
+    name: :class:`str`
+        The name of the POI.
+    location: :class:`fortnite_api.POILocation`
+        The location of the POI.
```
We might add somewhere the info about this being the location on the map specifically. To calculate the location on the image, they need to get the current map radius.
```py
x = (poi_y + world_radius) / (world_radius * 2) * map_width;
y = (1 - (poi_x + world_radius) / (world_radius * 2)) * map_height;
```

trevorflahardy replied:
Isn't this a very specific use case though? If a user wanted to paste the name of the POI onto the map why wouldn't they just request the POI image with labels from the API

Luc1412 replied:
I feel like it's a common use-case for map related stuff and might be a nice util. 

It might be used to mark specific POIS on the map.

trevorflahardy replied:
Sounds good to me, can you do the implementation? I wasn't aware of this until you brought it up so you have a better idea for function names and whatnot that I would 👍

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+    *   - ``client.cosmetics.search_all``
+        - :meth:`fortnite_api.Client.search_br_cosmetics`
+    *   - ``client.cosmetics.search_first``
+        - :meth:`fortnite_api.Client.search_br_cosmetics`
```
The old methods missing the brackets

Luc1412 commented on fortnite_api/new.py:
```diff
@@ -0,0 +1,224 @@
+        self.build: str = data['build']
+        self.previous_build: str = data['previousBuild']
+        self.date: datetime.datetime = parse_time(data['date'])
+        self.raw_data: Dict[str, Any] = data
```
Br New cosmetics is legacy, so can be removed including client fetch method, migration guide info

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+
+
+Of course, the same applies to the SyncClient client. The methods are the same, but they are synchronous and
+under the :class:`fortnite_api.SyncClient` client instead.
```
We could mention the deprecation of Creative News

Luc1412 commented on docs/migrating.rst:
```diff
@@ -0,0 +1,895 @@
+        - :class:`fortnite_api.BrInputStats`
+    *   - ``BrGameModeStats``
+        - :class:`fortnite_api.BrGameModeStats`
+
```
We could mention the deprecation of trios stats

Luc1412 commented on fortnite_api/material.py:
```diff
@@ -0,0 +1,140 @@
+
+    __slots__: Tuple[str, ...] = ('offer_image', 'background', 'texture')
+
+    def __init__(self, *, data: Dict[str, Any], http: HTTPClientT) -> None:
+        self.offer_image: Asset[HTTPClientT] = Asset(url=data['OfferImage'], http=http)
+
+        _texture = data.get('Texture')
+        self.texture: Optional[Asset[HTTPClientT]] = _texture and Asset(url=_texture, http=http)
+
+        _background = data.get('Background')
+        self.background: Optional[Asset[HTTPClientT]] = _background and Asset(url=_background, http=http)
```
We got a difficult spot here. These keys are dynamic. We could get a list of all existing keys, but in future updates, new can be added. Here a list of all current ones:
```py
{'FNMTexture', 'Flipbook', 'Graphic', 'Texture', 'WideTile', '2-OfferImage', 'OfferImage', 'ImageBackground', 'Winterfest_Red', 'NormalTile', 'SeriesTexture', 'Background'}
```

trevorflahardy replied:
All keys are going to resolve to `Assets`, so it may be best to just transform all keys into `Mapping[str, Asset]` then (except `offer_image`)? I believe that trying to keep the dynamic keys as attributes may lead to issues down the road.

Luc1412 replied:
Maybe we can type soft-type the dict. At least `OfferImage` and the manually supplied  Background could be more easily accessible, since they as used the most.

trevorflahardy replied:
Could you elaborate on what you mean by "type soft-type the dict"? I'm thinking that we basically transform the class to something like this:
```python
class MaterialInstanceImages(...):
    def __init__(self, *, data, http):
        self.offer_image: ... = ...
        self.raw: Dict[str, Asset[...]] = ...

    @property 
    def background(self) -> Optional[Asset[...]]: ...
 ```
 `offer_image` would be the only concrete attribute because it's manually supplied and the other keys that must be easily accessible will be properties. Additionally, `raw` will contain the parsed dynamic keys
 
 What do you think about this approach?

Luc1412 replied:
`OfferImage` isn't manually supplied, but it's the most useful out of all. `Background` is created and passed by Fortnite-API.com and contains the OfferImage with the generated background. Don't understand, that `background` is a prop, but `offer_image` a attribute.

Maybe we implement `__getitem__` for accessing other keys and providing the most common ones as an attribute or prop.

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,394 @@
+        ]
+        self.lego_kits: List[CosmeticLegoKit[HTTPClientT]] = [
+            CosmeticLegoKit(data=item, http=http) for item in get_with_fallback(data, 'legoKits', list)
+        ]
```
I think we should add a shortcut to iterate trough all items. If it's a `items` property that provides an iterable or a `__iter__` for `ShopEntry`

trevorflahardy replied:
`__iter__` for `ShopEntry` definitely made the most sense, added.

Luc1412 commented on fortnite_api/shop.py:
We need to add the 2 new attributes for the shop entry: `inDate` and `outDate` marking the dates the items rotate in/out the shop.

Luc1412 commented on fortnite_api/enums.py:
```diff
@@ -90,45 +263,114 @@ class BrCosmeticType(Enum):
     BATTLE_ROYALE = 'ECosmeticCompatibleMode::BattleRoyale'
     LEGO = 'ECosmeticCompatibleMode::Juno'
     ROCKET_RACING = 'ECosmeticCompatibleMode::DelMar'
     FESTIVAL = 'ECosmeticCompatibleMode::Sparks'
```
We got an issue. There are a new enum `ECosmeticCompatibleModeLegacy` values, which are identical to  `ECosmeticCompatibleMode`, except a `ECosmeticCompatibleModeLegacy::MAX` value which could be interpreted as `ALL`.

We might need to make this enum compatible with `ECosmeticCompatibleMode` and `ECosmeticCompatibleModeLegacy` + Adding an `ALL` value.

trevorflahardy committed:
Add pyright dependency to extras require

Luc1412 commented:
Am currently not sure about the design of creating classes. It's pretty cumbersome to re-create classes, since you always have to pass the HTTP. 

Not sure how much it's used, but I work a lot with caching, where I convert the data to json to drop it into redis e.g. `fortnite_api.Shop(data=json.loads(before_shop_data), http=self.bot.fortnite_api.http)`

However not sure if there is a better design.

Luc1412 commented:
I think it makes sense to use timezone-aware datetimes throughout the library. This is also a breaking change, that needs to be mentioned in the migration guide.

Luc1412 commented:
I think it makes sense to use timezone-aware datetimes throughout the library. This is also a breaking change, that needs to be mentioned in the migration guide.

Luc1412 commented:
Am currently not sure about the design of creating classes. It's pretty cumbersome to re-create classes, since you always have to pass the HTTP. 

Not sure how much it's used, but I work a lot with caching, where I convert the data to json to drop it into redis e.g. `fortnite_api.Shop(data=json.loads(before_shop_data), http=self.bot.fortnite_api.http)`

However not sure if there is a better design.

trevorflahardy committed:
Move ShopEntrty to ListTransformerProxy

trevorflahardy committed:
Add iter and len to shop entry

trevorflahardy committed:
Update tests to adjust for new shop updates

trevorflahardy committed:
Fix client name in outdated code block

trevorflahardy committed:
Remove external auth references from the codebase

trevorflahardy committed:
Correct doc for aes main key

trevorflahardy committed:
Transform map.pois to List[POI] and remove get_x function

trevorflahardy committed:
Black isort

trevorflahardy commented:
> Am currently not sure about the design of creating classes. It's pretty cumbersome to re-create classes, since you always have to pass the HTTP.

This is something that I've brought a lot of thought to. Ideally speaking `client.http` remains internal, so a helper function on the client may be best to recreate all objects that require a `http` parameter. Maybe,
```python
obj = client.reconstruct(ObjectType, data)
```
or some decorator that wraps each reconstructable object,
```python
@reconstructable
class Foo(...):
    ...

foo = Foo.reconstruct(client, data)
```
This can be typed well with `Protocol`s  also, so it still meets the library's type standards. What do you think?

Luc1412 commented:
> This is something that I've brought a lot of thought to. Ideally speaking `client.http` remains internal, so a helper function on the client may be best to recreate all objects that require a `http` parameter.

Maybe we rename `raw_data` to `to_dict` (or `as_dict`) and add `from_dict` (alternatively keep `raw_data` and add `from_data`). That `from_x`method got client param and the dict.

Luc1412 commented on fortnite_api/material.py:
```diff
@@ -0,0 +1,140 @@
+        self.background: Optional[Asset[HTTPClientT]] = _background and Asset(url=_background, http=http)
+
+
+class MaterialInstanceColors:
+    """
+    .. attributetable:: fortnite_api.MaterialInstanceColors
+
+    Represents some metadata about the colors of a Material instance. Every material instance image
+    has a background gradient, made up of a background color A and background color B that falls off to a
+    specific color. This class represents that information.
+
+    Attributes
+    ----------
+    background_color_a: Optional[:class:`str`]
+        The first background color of the Material instance, if any.
+    background_color_b: Optional[:class:`str`]
+        The second background color of the Material instance, if any.
+    fall_off_color: Optional[:class:`str`]
+        The fall off color of the Material instance, if any.
+    """
+
+    __slots__: Tuple[str, ...] = ('background_color_a', 'background_color_b', 'fall_off_color')
+
+    def __init__(self, *, data: Dict[str, Any]) -> None:
+        self.background_color_a: Optional[str] = data.get('Background_Color_A')
+        self.background_color_b: Optional[str] = data.get('Background_Color_B')
+        self.fall_off_color: Optional[str] = data.get('FallOff_Color')
```
Here we also got dynamic keys. Here we should also use a similar approach to `MaterialInstanceImages`

Luc1412 commented on fortnite_api/cosmetics/common.py:
```diff
@@ -0,0 +1,208 @@
+    def __init__(self, *, data: Dict[str, Any]) -> None:
+        self.value: CosmeticType = CosmeticType(data['value'])
+        self.display_value: str = data['displayValue']
+        self.backend_value: str = data['backendValue']
```
I'm currently not the biggest fan of the enum like value implementation. Currently it's `cosmetic.type.value.value` to get the value. Is there a way that we can implement this better?

trevorflahardy replied:
I noticed this while implementing the rewrite into my Discord bot, it is cumbersome. I'm torn on the solution for this though.

Conventionally this attr should remain as an enum because it's what the end user is expected to use. This, however, does not get around the issue of obtaining the underlying raw value, which is what many users are going to use. Perhaps a `raw_value` (`str`) attribute should be added for this reason? What do you think

trevorflahardy replied:
Although we discussed potentially reworking the `CosmeticType` enum, this ultimately is not going to work. Not only can an enum not be extended, it could lead to confusion because the enum is also used as some parameters on other API calls. I've exposed the value through a `raw_value` attribute, such that:
```python
val = cosmetic.type.raw_value
```

Luc1412 commented on fortnite_api/cosmetics/instrument.py:
```diff
@@ -0,0 +1,117 @@
+        ]
+
+    @property
+    def showcase_video(self) -> Optional[str]:
```
adding `_url` subfix, might make it more clear that it's not the video itself, but the url.

Luc1412 commented on fortnite_api/cosmetics/common.py:
```diff
@@ -0,0 +1,208 @@
+
+    __slots__: Tuple[str, ...] = ('featured', 'lego', 'other')
+
+    def __init__(
+        self,
+        *,
+        data: Dict[str, Any],
+        http: HTTPClientT,
+    ) -> None:
+        super().__init__(data=data, http=http)
+
+        featured = data.get('featured')
+        self.featured: Optional[Asset[HTTPClientT]] = featured and Asset(http=http, url=featured)
+
+        lego = data.get('lego')
+        self.lego: Optional[Asset[HTTPClientT]] = lego and Asset(http=http, url=lego)
+
+        # Mapping of str to Asset. In testing all I have found is "background" in this mapping.
+        other: Dict[str, str] = get_with_fallback(data, 'other', dict)
```
This misses many stuff.

`CosmeticBr` usually got `smallIcon`, `icon`, `featured`, `lego` and `other`
`CosmeticInstrument`, `CosmeticCar` and `CosmeticLegoKit` usually got `small` and `large`
`CosmeticLego` usually got `small`, `large`, `wide`

trevorflahardy committed:
Implement repr(x) for all objects

Luc1412 commented on fortnite_api/cosmetics/lego.py:
```diff
@@ -0,0 +1,120 @@
+        self.wide: Optional[Asset[HTTPClientT]] = _wide and Asset(http=http, url=_wide)
+
+
+class CosmeticLego(Cosmetic[HTTPClientT]):
```
Not sure about the naming, since it's not a real cosmetic, rather a Lego Version of an existing cosmetic.

trevorflahardy replied:
Regardless of where the cosmetic stems from it is nonetheless a cosmetic because it's a usable item in Fortnite Lego, no?

Luc1412 replied:
I mean you buy a specific skin in the Shop. If you play normal modes, you play the normal variant. If you play LEGO modes, the lego variant. This CosmeticLego, describes this LEGO variant.

trevorflahardy replied:
Just `Lego` then perhaps? This would require the class to be moved out of `/cosmetics` and all the docs to be updated too if the API does not consider it a cosmetic

Luc1412 replied:
I think Lego isn't a good name. Rather `LegoVariant`.

trevorflahardy committed:
Document repr

trevorflahardy committed:
Correct dataclass usage

trevorflahardy committed:
Fix key error

trevorflahardy committed:
Add missing simple repr

trevorflahardy committed:
Correct key error on map

trevorflahardy committed:
Update tests for poi

trevorflahardy committed:
Update tests for POI

trevorflahardy committed:
Correct documentation from reveal_type to print

trevorflahardy committed:
Add the other keys as properties on CosmeticImages

trevorflahardy committed:
Replace showcase video with urk postfix

trevorflahardy committed:
Upgrade cosmetic images to blanket all cosmetic types

trevorflahardy committed:
Remove CosmeticLegoImages and move under blanket

trevorflahardy committed:
Add missing brackets

trevorflahardy committed:
Update documentation for client parameters

trevorflahardy committed:
Add note for datetime objects

trevorflahardy committed:
Black isort

trevorflahardy committed:
Correct documentation for BrGameModeStats last modified

trevorflahardy committed:
Remove mythic rarity

trevorflahardy committed:
Remove all references to NewBrCosmetics

trevorflahardy committed:
Note that trios is not available.

trevorflahardy committed:
Remove all Optmization Flags

trevorflahardy committed:
Resolve issue with `cosmetic.type.value.value`. Change to `cosmetic.type.raw_value`

trevorflahardy committed:
Make MaterialInstanceImages and MaterialInstanceColors dynamic in nature

trevorflahardy committed:
Update formatting on async methods tests

trevorflahardy committed:
Move to pyproject.toml instead of setup.py

trevorflahardy committed:
Remove Medium intensity

trevorflahardy committed:
Update authors

trevorflahardy committed:
Impl in_date and out_date on ShopEntry

trevorflahardy committed:
Correct README.md formatting

trevorflahardy committed:
Beef README.md

trevorflahardy committed:
Add mythic cosmetic rarity

trevorflahardy committed:
Update banner class (for issue) and fix finalize TODO note

trevorflahardy committed:
Update test for account_repr

trevorflahardy committed:
refactor: Update primary mode parsing in MaterialInstance constructor

trevorflahardy committed:
Update tests

trevorflahardy committed:
Update tests for TileSize checks

trevorflahardy committed:
Black isort

trevorflahardy committed:
Doc link fixes

trevorflahardy committed:
- Rename CosmeticLego -> VariantLego
- Move VariantLego to `cosmetics/variants`

trevorflahardy committed:
Black isort, add bean file

trevorflahardy committed:
Implement VariantBean, cleanup docs

trevorflahardy committed:
Add bean cosmetic onto new response

trevorflahardy committed:
Update docs for Variants

trevorflahardy committed:
Fixes for new Bean

trevorflahardy committed:
Document enum for docs building

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -1,266 +1,463 @@
+# Matches a tile size: size_<int>_x_<int>
+_TILE_SIZE_REGEX = re.compile(r"size_(?P<width>\d+)_x_(?P<height>\d+)", re.IGNORECASE)
+
+_BACKUP_TILE_MAPPING: Mapping[str, Tuple[int, int]] = {
+    "mini": (1, 1),
+    "small": (1, 1),
+    "normal": (1, 2),
+    "doublewide": (2, 2),
+    "triplewide": (3, 2),
+    "quadwide": (4, 2),
+    "pentawide": (5, 2),
+}
```
I think we can remove the backup tile mapping since Epic only uses the new sizing

Luc1412 committed:
Merge branch 'master' into rewrite

trevorflahardy committed:
Remove tile size backup, raise exception on failed regex

trevorflahardy committed:
Add docs to TileSize and update docs

trevorflahardy committed:
Add Reconstructable and new properties

trevorflahardy committed:
Add reconstructable to all base classes

trevorflahardy committed:
Impl reconstructable for common cosmetics

trevorflahardy committed:
Impl reconstructable for br cosmetics

trevorflahardy committed:
Widen the type of Mapping in DictT

trevorflahardy committed:
fix impl for car with reconstructable

trevorflahardy committed:
impl Reconstructable for all base cosmetics

trevorflahardy committed:
Add support for reconstructable for lego and bean

trevorflahardy committed:
Fix TypeVar type

trevorflahardy committed:
Update priv variable use

trevorflahardy committed:
Update tests for new changes

trevorflahardy committed:
Update check for from_dict

trevorflahardy committed:
Add loose test for reconstruction

trevorflahardy committed:
Black isort

trevorflahardy committed:
Update docs references, add ReconstructAble to docs

trevorflahardy committed:
Remove uncaught references and inherits related to raw_data attrs

trevorflahardy committed:
Documentation changes to migrating

trevorflahardy committed:
Doc changes to index (main page)

trevorflahardy committed:
Doc changes for formatting

trevorflahardy committed:
Black isort

trevorflahardy committed:
Remove usage of pop in MaterialInstanceImages

trevorflahardy committed:
Refactor __repr__ method in utils.py to exclude 'raw_data' attribute from representation

trevorflahardy committed:
Add examples, rename from stats to input

trevorflahardy committed:
Update spacing

trevorflahardy committed:
Refactor list-table in migrating.rst for consistent formatting

trevorflahardy committed:
Refactor NewCosmetics class to improve readability and maintainability

trevorflahardy committed:
black isort

trevorflahardy committed:
Update packages to include the following:
- Module cosmetics
- Module cosmetics.variants

trevorflahardy committed:
Remove empty duplicate test file

trevorflahardy committed:
Correct len() issue with beans

trevorflahardy committed:
black isort

trevorflahardy committed:
Upgrade discord integration example

trevorflahardy committed:
Update docs for examples

trevorflahardy committed:
Update and fix for Material instance key error.

trevorflahardy committed:
Fix to use API key in beta fetch material instances tests

Luc1412 committed:
Merge branch 'master' into rewrite

trevorflahardy committed:
Add Response Flags

trevorflahardy committed:
Add ResponseFlags to docs

trevorflahardy committed:
Merge branch 'rewrite' of https://github.com/Fortnite-API/py-wrapper into rewrite
Merge branches

trevorflahardy committed:
Black isort, moving onto the Client impl of Flags

trevorflahardy committed:
Add pre commit hooks for black and isort

trevorflahardy committed:
Update dev dependencies for pre commit

trevorflahardy committed:
Add all to ResponseFlags

trevorflahardy committed:
Add an all() classmethod to ResponseFlags class

trevorflahardy committed:
Add response flags to all relevant HTTP methods

trevorflahardy committed:
Add flags export to init

trevorflahardy committed:
Update client for ResponseFlags:
- Add init parameter and document changes
- Add new method to parse any passed response flags
- Add new response flags parameters to all relevant methods and document changes

Update to Client:
- Change how passing of `language=None` is handled. Now, specifically passing `None` will not set the language parameter in the request. This is to allow for the language parameter to be set to `None` in the request if the user wishes.

trevorflahardy committed:
Black and isort

trevorflahardy committed:
Sync Client and AsyncClient for recent changes

trevorflahardy committed:
Correct for doc fixes

trevorflahardy committed:
Add flags to tests

trevorflahardy committed:
Fix KeyError when a field is None in hashes

trevorflahardy committed:
Updates to how client parses in search_br_cosmetics

trevorflahardy committed:
Correct issue with flags not parsing correctly.

Update documentation for default_language field and make the field not optional due to its default.

Luc1412 committed:
Add response_flags to shop methods

Luc1412 committed:
Add series to CosmeticLegoKit

Luc1412 committed:
Cleanup imports

Luc1412 committed:
Improve grammar and typos

Luc1412 committed:
Improve and simplify css files

Luc1412 committed:
Improve situation shadowing of built-in methods

Luc1412 committed:
Add bean to CosmeticImages
Remove outdated documentation for other attribute

Luc1412 committed:
Improve situation shadowing of built-in methods

Luc1412 committed:
Fix issue with docs for GameLanguage.PORTUGUESE_BRASIL

Luc1412 committed:
Improve code regarding dict creation and comparisons

Luc1412 committed:
Remove unnecessary overwriting of small_icon and icon in CosmeticImages
